### PR TITLE
#11211 Fix Store View switcher

### PR DESCRIPTION
**Pull request by thiagolima-bm**
Store View switcher not working on front-end and it throws an error #11211 #10908

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=11337) <br/>All committers have signed the CLA.

File: app/code/Magento/Store/Model/Store.php [hunk 1]
```diff
@@ -1166,6 +1166,9 @@ public function getCurrentUrl($fromStore = true)
         if (!$this->isUseStoreInUrl()) {
             $storeParsedQuery['___store'] = $this->getCode();
         }
+        if ($this->getCode() !== $this->_storeManager->getStore()->getCode()) {
```
Comment (vkublytskyi) [inline app/code/Magento/Store/Model/Store.php]: @thiagolima-bm Please explain why usage of `__from_store` query parameter should be used.

File: app/code/Magento/Store/Model/Store.php [hunk 2]
```diff
@@ -1175,9 +1178,14 @@ public function getCurrentUrl($fromStore = true)
...
             . $storeParsedUrl['host']
             . (isset($storeParsedUrl['port']) ? ':' . $storeParsedUrl['port'] : '')
-            . $storeParsedUrl['path']
-            . $requestString
-            . ($storeParsedQuery ? '?' . http_build_query($storeParsedQuery, '', '&amp;') : '');
+            . $storeParsedUrl['path'];
+
+        //avoid query params duplication
+        if (!preg_match('/___store=(.*?)&___from_store=(.*?)/', $requestString)) {
+            $currentUrl .= $requestString;
```
Comment (vkublytskyi) [inline app/code/Magento/Store/Model/Store.php]: `$requestCode` also contains path and this data may be lost

STR:
- go to Privacy and Cookie Policy
- switch store view (store view switched, user still on Privacy and Cookie Policy)
- switch store view again (store view switched, user on main page but should be on Privacy and Cookie Policy)

File: app/code/Magento/Store/Model/Store.php [hunk 3]
```diff
@@ -1175,9 +1178,14 @@ public function getCurrentUrl($fromStore = true)
...
+        }
+
+        $currentUrl .= ($storeParsedQuery ? '?' . http_build_query($storeParsedQuery, '', '&amp;') : '');
```
Comment (vkublytskyi) [inline app/code/Magento/Store/Model/Store.php]: I think the source of the issue is concatenation with `?` while `$requestString` may contain query. To fix this I propose:
1. Split `$requestString` into path and query:
```php
$requestStringParts = explode('?', $requestString, 2);
$requestStringPath = $requestStringParts[0];
if (isset($requestStringParts[1])) {
   parse_string($requestStringParts[1], $requestStringQuery);
} else {
    $requestStringQuery = []
}
```
2. Merge original query params and store parsed query
```php
$currentUrlQueryParams = array_merge($requestStringQuery, $storeParsedQuery);
```
3. Safely concatenate all parts of url:
```php
$currentUrl = $storeParsedUrl['scheme']
    . '://'
    . $storeParsedUrl['host']
    . (isset($storeParsedUrl['port']) ? ':' . $storeParsedUrl['port'] : '')
    . $storeParsedUrl['path']
    . $requestStringPath
    . ($currentUrlQueryParams ? '?' . http_build_query($currentUrlQueryParams, '', '&amp;') : '');
```

**Review by vkublytskyi — CHANGES_REQUESTED** [review]

File: app/code/Magento/Store/Model/Store.php (hunk 1 above)
Comment (thiagolima-bm) [reply app/code/Magento/Store/Model/Store.php]: for tracking, maybe?

**Comment by thiagolima-bm** [comment]
@vkublytskyi fixes applied.

**Review by vkublytskyi — APPROVED** [review]

**Comment by chiragit007** [comment]
45

**Comment by rsulym** [comment]
Gents, still not able to manage this issue to get it work ((
Version 2.2.1, cannot upgrade to 2.2.2
Is this file (Store.php) is complete fix which i need?
https://github.com/thiagolima-bm/magento2/blob/84729b581638e00459fcca63bb6d7a1c77f252b9/app/code/Magento/Store/Model/Store.php

**Pull request merged**